### DISCUSSION TITLE
socket emitting same data to different clients

### DISCUSSION THREAD

**Discussion by vrajeshbhatt** [miguelgrinberg/Flask-SocketIO#1786] [Q&A]
Hello there,
In my project, I am using flask socketio as a server and socket io for the client (javascript).
In flask, there is a function for sending the data to client which is connected with sid. The issue I am facing is when multiple user/client connects the data which is the server/flask function is sending is somehow mixing up.

for example, if user A and B are there and accessing the same exercise page (javascript client) so if user A do some exercise which leads to some result that comes from the server (flask socketio) is reflected on both side client A and B.
Even I am checking the session id with the conditions with the request.sid in server-side and socket.id in client-side. But the issue is the same.

### **client side code**
```
socket.emit("high_knee_landmarks", results.poseLandmarks,socket.id);
socket.emit("high_knee_data",socket.id);
socket.on("recv_data", function (data) {
        if (socket.id == data.id){
          
           document.getElementById("rep_count").innerHTML = data.rep_count;
          ........}
```

### **server side code**
```
@socketio.on('high_knee_landmarks')
def high_knee_landmarks(data,id):
   
    if id == request.sid:
        global high_knee_pose_row
        high_knee_pose_row = list(np.array([[landmark['x'], landmark['y'], landmark['z'], landmark['visibility']] for landmark in data]).flatten())

@socketio.on('high_knee_data')
def high_knee_data(id):
    if id == request.sid:
        previous_stage,points= high_knee(high_knee_pose_row,previous_stage_high_knee)
        emit('recv_data',{'points':high_knee_points,'rep_count':rep_count,'id':request.sid})
```
I also tried
`socket.emit('recv_data',{'points':high_knee_points,'rep_count':rep_count,'id':request.sid})
`and 
`emit('recv_data',{'points':high_knee_points,'rep_count':rep_count,'id':request.sid},to=request.sid)
` 
But still facing the same issue.

I will appreciate it if you tell me what is the problem and some solutions for the same.

**Comment by miguelgrinberg** [comment]
You are using global variables. How does that work when you have two or more clients?

**Reply by vrajeshbhatt** [reply]
when a new user comes the global variables are re-initialized to their original state.

**Reply by miguelgrinberg** [reply]
Yes. And that destroys the variables for the first client. You can't handle more than one client at a time with your solution. So it is expected that you will have issues when two clients are accessing your application at the same time.

**Reply by vrajeshbhatt** [reply]
Thank you for the reply.
Now, I understand what's happening. But I am using a global variable to access the variable of another function. 
Is there a way around handling multiple clients?

**Reply by miguelgrinberg** [reply]
There's many ways. Which one's the best really depends. One option is to use a single function for everything. The client emits, and on the server you do your calculations and emit when you have a result, without having to create a separate interaction for the results. Another option is to use a global dictionary, where you store results for multiple clients under the `request.sid` key.

**Reply by vrajeshbhatt** [reply]
Thank you sir for the suggestion. I will try to work with that.
